### PR TITLE
Add action "revoke sessions" to users_controllers

### PR DESCRIPTION
**Pull request by lpichler**
Add endpoint to revoke user's session as described here in this issue: https://github.com/ManageIQ/manageiq-api/issues/884

Current Signature:
https://github.com/ManageIQ/manageiq-api/pull/900#issuecomment-707271568

## Links
Issue - https://github.com/ManageIQ/manageiq-api/issues/884 
Parent Issue - https://github.com/ManageIQ/manageiq/issues/20378
Next Issue (follow up) - https://github.com/ManageIQ/manageiq-api/issues/885

## Needs to be merged before (reason for WIP)
- [x] https://github.com/ManageIQ/manageiq/pull/20601
- [x] https://github.com/ManageIQ/manageiq/pull/20633
- [x] https://github.com/ManageIQ/manageiq/pull/20692

### CODE REVIEW

**Comment by lpichler** [comment]
cc @gtanzillo @NickLaMuro

File: app/controllers/api/revoke_controller.rb
Comment (abellotti) [inline app/controllers/api/revoke_controller.rb:14]: we might also need to check if target user is not the current user, then current_user.super_admin_user? should be true unless we introduce a new entitlement for this, in which case something liek current_user.role_allows?(:identifier => 'revoke_user_sessions') could be introduced.

File: app/controllers/api/revoke_controller.rb
Comment (gtanzillo) [inline app/controllers/api/revoke_controller.rb:18]: I'm thinking that the current user should be able to revoke his own sessions regardless of RBAC, just like he'd be able to log himself out. Do you think we should skip RBAR for this case?

I'm also thinking this would be the most common way this API will be accessed

File: app/controllers/api/revoke_controller.rb
Comment (lpichler) [reply app/controllers/api/revoke_controller.rb]: @abellotti I updated PR as you suggested:
- I added check if target user is not the current user 
- I added feature `revoke_user_sessions` in core for this operation
- I refactored it a bit.

can you take a look please ? thanks

File: app/controllers/api/revoke_controller.rb
Comment (lpichler) [reply app/controllers/api/revoke_controller.rb]: yes I think also that we don't need to do RBAC check on current user. 

Newly I added here `role_allows?` check for product feature `revoke_current_user_sessions` and
thanks to that we can distinguish privileges
for revoking current user's sessions and 
for revoking sessions of requester users from params. 

thanks

File: app/controllers/api/revoke_controller.rb
Comment (abellotti) [inline app/controllers/api/revoke_controller.rb:37]: I just want to make sure of the following:
- A user can revoke his own sessions with or without the userid param and without needing the revoke_user_sessions entitlement.
- A user can revoke someone else's session if and only if they have the revoke_user_sessions entitlement.

Comment (lpichler) [reply app/controllers/api/revoke_controller.rb:37]: I think it is not possible so simply as you suggesting and we need to use more entitlements.

- we need one entitlement(product feature added [here](https://github.com/ManageIQ/manageiq/pull/20633) 1- `revoke_session`) just to allow access method `update` -  this is done generally and this check is done in https://github.com/ManageIQ/manageiq-api/blob/master/app/controllers/api/base_controller/parser.rb#L178.
It finds entitlement in api.yml(added in this PR ) for edit action and then it is checked whether current user has this entitlement(product feature added)
- > - A user can revoke his own sessions with or without the userid param and without needing the revoke_user_sessions entitlement.

We are not able to revoke without `revoke_user_sessions ` but we need also  entitlement `revoke_session` to have on current user from first point and thanks to that we need to add new entitlement just for this case so I added `revoke_current_user_sessions`.

- > - A user can revoke someone else's session if and only if they have the revoke_user_sessions entitlement.

Yes this is correct, this operation can be performed only if current (authenticated) user has `revoke_user_sessions` entitlement(product feature) and if current (authenticated) user has RBAC access to this user.


Hopefully this is more clear and it makes sense now.
If you have any other concerns/suggestions please let me know.

thanks

**Review by gtanzillo — APPROVED** [review]

File: app/controllers/api/revoke_controller.rb
Comment (abellotti) [reply app/controllers/api/revoke_controller.rb:37]: I wonder if you need to update https://github.com/ManageIQ/manageiq/blob/master/db/fixtures/miq_user_roles.yml  in the https://github.com/ManageIQ/manageiq/pull/20633 PR so you add the `revoke_user_sessions` to let's say admins and super-admins, and add `revoke_sessions` to every other role.

Comment (lpichler) [reply app/controllers/api/revoke_controller.rb]: yes that's a good note, I did update -
`super_administrator` and `administrator` 
have privileges(`revoke_sessions` as parent entitlement) to perform both revoke operations and
other roles have privileges(`revoke_current_user_sessions`) to perform **only** revoke own session of authenticated (current)user. 

thanks

**Comment by gtanzillo** [comment]
Regarding 
>A user can revoke his own sessions with or without the userid param and without needing the revoke_user_sessions entitlement.

Couldn't we just change the method to do `||` for the three conditions? Like -
```
    def current_user_allowed?(userid)
      (userid.nil? || userid == current_user.userid) || current_user.role_allows?(:identifier => 'revoke_current_user_sessions')
    end
```

**Comment by lpichler** [comment]
> Regarding
> 
> > A user can revoke his own sessions with or without the userid param and without needing the revoke_user_sessions entitlement.
> 

@gtanzillo yes we could but this statement is not reachable because we need one product feature to access  `update` which is done by common code. I described it more [here](https://github.com/ManageIQ/manageiq-api/pull/900#discussion_r498455085)

I used that one feature as parent feature for other features which are tied with operations:
1. revoke own sessions of authenticated (current)user
2. revoke sessions of requested user from parameters

then the mentioned method have different semantic - method checks requirement on `revoke_current_user_sessions` for current user but only when **case 1. is valid** (no `userid` is present in request parameters - and yes there is one exception when passed `userid` is same asi current user)


let me know if have more questions, thanks

**Comment by lpichler** [comment]
@gtanzillo @abellotti thanks for today's discussion and  I updated PR as suggested:

- the only new entitlement `revoke_user_sessions` has added [here](https://github.com/ManageIQ/manageiq/pull/20633) 
- super admin a users with role `EvmRole-administrator` can revoke own's sessions and other's user sessions
- other users can revoke only own's sessions by default  (if they want to revoke other user's session they have to have enabled  permission and access to user regard to our rbac filterer)

PR ready for review again 👍

File: config/api.yml
Comment (Fryguy) [inline config/api.yml:3049]: I find this weird...I would expect a revoke action on the /users/ endpoint.

Comment (Fryguy) [reply config/api.yml:3049]: Either that or a revoke action on the /auth endpoint.

**Comment by Fryguy** [comment]
Code wise, everything looks good to me.  Ergonomics-wise I would never expect to hit a revokes endpoint when this is a function/action of either users or auth.

File: app/controllers/api/revokes_controller.rb
Comment (Fryguy) [inline app/controllers/api/revokes_controller.rb:5]: Most of these methods should be made private except the ones exposed to routes.

File: config/api.yml
Comment (lpichler) [reply config/api.yml:3049]: yes `/user` sounds better to me and based on some api suprises with current way `/user` fits better to our api.

it is based on @abellotti [suggestion](https://github.com/ManageIQ/manageiq-api/issues/884#issuecomment-67668246), 

@abellotti it is ok If I will change it to `/user` endpoint ?

thanks

**Comment by abellotti** [comment]
Since we can revoke ui session too for the user, I tend to lean toward the users endpoint since /api/auth is intended for the API session token.  Since we're not required to specify a user id, but target the authenticated user, we can leverage collection level action, i.e. https://github.com/ManageIQ/manageiq-api/pull/629


Maybe introducing a "revoke_sessions" action as follows:

```
POST /api/users
{
   "action" : "revoke_sessions",
   "session_types" : [ "ui", "api" ]
}
```

This will target the authenticated user,

But also support the resource level action:

```
POST /api/users/:id
{
   "action" : "revoke_sessions",
   "session_types" : [ "ui", "api", "ws" ]
}
```

And nothing stops us from supporting the collection level (bulk) action for revoking sessions of multiple users:

```
POST /api/users
{
   "action" : "revoke_sessions",
   "resources" : [
       { "id" : ":id1", "session_types" : ["ui"] },
       { "id" : ":id2", "session_types" : ["api"] },
       { "id" : ":id3" },
       { "id" : ":id4", "session_types" : ["ui", "api"]}
   ]
}
```

**Comment by Fryguy** [comment]
@abellotti Yes that is exactly what I was thinking... I just realized I put /api/revoke in the original requirements and that was my mistake...I actually meant to write something like `/api/user (action=revoke)`

**Comment by Fryguy** [comment]
I could also see an revoke_session (singular) action on the `api/user/<id>` endpoint to revoke the session for a specific user.

**Comment by lpichler** [comment]
@abellotti @gtanzillo @Fryguy I did changes according to signature in https://github.com/ManageIQ/manageiq-api/pull/900#issuecomment-707271568.

File: app/controllers/api/users_controller.rb
Comment (Fryguy) [inline app/controllers/api/users_controller.rb:69]: Should these 2 methods be private since they aren't routable?

**Review by Fryguy — APPROVED** [review]
LGTM - will defer to @abellotti for merge.

File: spec/requests/users_spec.rb
Comment (abellotti) [inline spec/requests/users_spec.rb:736]: minor -> doesn't

File: spec/requests/users_spec.rb
Comment (abellotti) [inline spec/requests/users_spec.rb:742]: Nice to see the bulk action working 👍

File: spec/requests/users_spec.rb
Comment (abellotti) [inline spec/requests/users_spec.rb:648]: minor -> other than ...

File: app/controllers/api/users_controller.rb
Comment (abellotti) [inline app/controllers/api/users_controller.rb:90]: I like how you're checking entitlement without needing the role entitlement in the api.yaml.  👍

**Comment by abellotti** [comment]
@lpichler this looks great and straightforward implementation.  Just minor typos, but other than that LGTM!!

File: app/controllers/api/users_controller.rb
Comment (lpichler) [reply app/controllers/api/users_controller.rb:69]: those method could be probably private but we don't have such methods set as private in other controllers, so let's leave as it is for now, maybe it is intentionally for some reason.

**Comment by lpichler** [comment]
@abellotti fix typos,
@Fryguy @abellotti for review 👍

**Review by abellotti — APPROVED** [review]

**Comment by abellotti** [comment]
LGTM @lpichler !!  Thanks for the API Enhancement. will merge when green.

**Pull request merged**
